### ISSUE TITLE
extra terminal processes hanging around on restart

### ISSUE THREAD

**Issue by meganrogge** [microsoft/vscode#174956]
Testing #174837

Regardless of `window.experimental.sharedProcessUseUtilityProcess`, for every `N` terminals in a window, and `R` restarts, there are `N * R` processes shown in the activity monitor.

https://user-images.githubusercontent.com/29464607/220416103-e771334a-5df1-44d9-bbf4-630ce5413fa0.mov

**Comment by meganrogge** [comment]
quit and reopened again and now there are 6

<img width="1729" alt="Screenshot 2023-02-21 at 11 24 51 AM" src="https://user-images.githubusercontent.com/29464607/220416499-e1e02ebb-8372-4dfb-ae3e-93813e56718f.png">

**Comment by meganrogge** [comment]
other info:

<img width="476" alt="Screenshot 2023-02-21 at 11 29 41 AM" src="https://user-images.githubusercontent.com/29464607/220417577-fa491577-c999-4fa1-81a8-bb043bb0e149.png">

**Comment by meganrogge** [comment]
They're not listed as zombie processes, so I believe they're still running. Reloading neither creates nor destroys processes

![Screenshot 2023-02-21 at 2 09 58 PM](https://user-images.githubusercontent.com/29464607/220447877-4554c3cc-10a9-438c-b98c-95dd37a0f209.png)

**Comment by meganrogge** [comment]
I have to select `Force Quit` to kill them

**Comment by rzhao271** [comment]
Similar issue on Linux.
On Ubuntu 22.04 on macOS Parallels, the shell integration scripts remain after closing code-insiders: `/usr/bin/bash --init-file /usr/share/code-insiders/resources/app/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh`. Using `ps aux | grep code-insiders`, the status of those processes is Ss+. I'm wondering whether something might be causing the shell integration scripts to fail?

**Comment by meganrogge** [comment]
interesting finding @rzhao271 . I wanted to see if this reproduces without shell integration enabled, but it looks like only terminals with shell integration persist at this pt, which is also unexpected and a different bug

**Comment by meganrogge** [comment]
also not reproducing with zsh anymore, though I'm sure it's still an issue because when I initially opened activity monitor (with zero terminals), I had 50+ zsh processes reported

**Comment by joaomoreno** [comment]
I can easily reproduce this in a vanilla Ubuntu using bash. The problem is the shell integration shell traps signals and ignores SIGTERM. The SIGKILL signal works just fine. https://github.com/microsoft/vscode/issues/176549

**Comment by dinofx** [comment]
The `freeze-slow-crash-leak` label might be an understatement?  Any processes that were started from the terminal continue to run after VSCode exits.

**Comment by Tyriar** [comment]
Repro out of sources on mac:

1. ./scripts/code.sh
2. Enable shell integration
3. Open a terminal
4. `ps | grep bash`, count procs
5. ctrl+c ./scripts/code.sh (doesn't repro on cmd+q)
6. `ps | grep bash`, :bug: count is the same

**Comment by Tyriar** [comment]
Can repro using this profile in 1.76.2:

```json
  "terminal.integrated.profiles.osx": {
    "bash": {
      "path": "bash",
      "args": [
        "--init-file",
        // Just an empty file
        "/Users/daimms/dev/microsoft/vscode/src/vs/workbench/contrib/terminal/browser/media/empty.sh"
      ],
      "icon": "terminal-bash"
    }
  }
```

Could this be related to the posix spawn change in node-pty?

**Comment by Tyriar** [comment]
The posix spawn change and other node-pty changes came in on Jan 19: https://github.com/microsoft/vscode/pull/171583

**Comment by meganrogge** [comment]
Yes, there's this too https://github.com/microsoft/vscode/issues/177044

**Comment by Tyriar** [comment]
Better repro for OSS: `kill -s SIGKILL <mainpid>`

**Comment by Tyriar** [comment]
This can be reproduced in OSS with:

1. ./scripts/code-cli.sh
2. Enable shell integration
3. Open a terminal
4. ps | grep bash, count procs
5. Open VS Code process explorer, get main process PID
6. kill -s SIGKILL <main process PID>
7. ps | grep bash, 🐛 count is the same

After downgrading node-pty, it **does not** reproduce:

```
yarn add node-pty@0.11.0-beta11
```

@deepak1556 @rzhao271 I'm going to need some help on this one, something between node-pty@0.11.0-beta12 and node-pty@0.11.0-beta30 caused this problem. Unfortunately it's not really possible to tell the exact commits that each beta build points at, the bundles are fairly small though. My guess is it's because of https://github.com/microsoft/node-pty/pull/487 since that blocked me from upgrading for over a year.

**Comment by deepak1556** [comment]
@Tyriar I can investigate this today.

**Comment by rsodre** [comment]
The issue doesn't seem to be resolved in version 1.77, as stated in #176819

**Comment by Varun-garg** [comment]
> Similar issue on Linux. On Ubuntu 22.04 on macOS Parallels, the shell integration scripts remain after closing code-insiders: `/usr/bin/bash --init-file /usr/share/code-insiders/resources/app/out/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh`. Using `ps aux | grep code-insiders`, the status of those processes is `Ss+` (waiting for an event to complete, and is a "session leader in the foreground process group"). I'm wondering whether something might be causing the shell integration scripts to fail to complete?

On gnome this also causes lag during the shutdown. Stale `shellIntegration-bash.sh`, also ignores SIGTERM sent during shutdown, which then waits for two minutes and finally sends SIGKILL

**Comment by joshstrike** [comment]
Per #180894 :
On MacOS the processes are properly killed when explicitly closing the workspace window, but remain running if you simply quit VSCode.

This behavior is the same even with `terminal.integrated.enablePersistentSessions` set to false. It is the same in all three modes of `terminal.integrated.persistentSessionReviveProcess`.

**Comment by meganrogge** [comment]
@joshstrike thanks for that info. by "quit VS code" I assume you mean via the mac dock?

![Screenshot 2023-04-27 at 1 37 09 PM](https://user-images.githubusercontent.com/29464607/234985034-98d6da2f-0444-4321-b672-af3b9c9ed82a.png)

**Comment by joshstrike** [comment]
@meganrogge 
Quitting via Cmd-Q. But the same happens when quitting via the dock.
Cmd-Shift-W to close the window terminates all running shells.

**Comment by deepak1556** [comment]
@joshstrike does the issue repro on insiders as well https://code.visualstudio.com/insiders ?

**Comment by joshstrike** [comment]
@deepak1556 Confirm, insiders version fixes the issue, all terminals correctly close upon Cmd-Q quit.

**Comment by deepak1556** [comment]
Thanks for confirming, the fix will be available with `v1.78` next week.
